
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Accessing curated gene expression data with GemmaPy &#8212; GemmaPy 0.0.1 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="API Reference" href="api.html" />
    <link rel="prev" title="gemmapy: a Python Wrapper for the Gemma API" href="index.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="accessing-curated-gene-expression-data-with-gemmapy">
<h1>Accessing curated gene expression data with GemmaPy<a class="headerlink" href="#accessing-curated-gene-expression-data-with-gemmapy" title="Permalink to this heading">¶</a></h1>
<section id="about-gemma">
<h2>About Gemma<a class="headerlink" href="#about-gemma" title="Permalink to this heading">¶</a></h2>
<p><a class="reference external" href="https://gemma.msl.ubc.ca/">Gemma</a> is a web site, database and a set
of tools for the meta-analysis, re-use and sharing of genomics data,
currently primarily targeted at the analysis of gene expression
profiles.  Gemma contains data from thousands of public studies,
referencing thousands of published papers.  Every dataset in Gemma has
passed a rigorous curation process that re-annotates the expression
platform at the sequence level, which allows for more consistent
cross-platform comparisons and meta-analyses.</p>
<p>For detailed information on the curation process, read this <a class="reference external" href="https://pavlidislab.github.io/Gemma/curation.html">page</a> or the latest
<a class="reference external" href="https://academic.oup.com/database/article/doi/10.1093/database/baab006/6143045">publication</a>.</p>
</section>
<section id="installation-instructions">
<h2>Installation instructions<a class="headerlink" href="#installation-instructions" title="Permalink to this heading">¶</a></h2>
<p>The package has been tested with Python3.7 but it should work with any Python3 version.</p>
<p>Install it with:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>pip install -i https://test.pypi.org/simple/ --no-deps gemmapy
<span class="c1"># install dependencies</span>
pip install urllib3 setuptools certifi anndata six python-dateutil pandas numpy
</pre></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The installation command will change as the first release of the package
is uploaded to the PyPI repository (vs TestPyPI now).</p>
</div>
</section>
<section id="additional-packages">
<h2>Additional packages<a class="headerlink" href="#additional-packages" title="Permalink to this heading">¶</a></h2>
<p>For the purpose of making plots in this tutorial the following packages should be installed
and imported: <code class="code docutils literal notranslate"><span class="pre">matplotlib</span></code>, <code class="code docutils literal notranslate"><span class="pre">plotnine</span></code>.</p>
</section>
<section id="downloading-expression-data">
<h2>Downloading expression data<a class="headerlink" href="#downloading-expression-data" title="Permalink to this heading">¶</a></h2>
<p>The main goal of this wrapper is to enable easy access to Gemma’s curated
datasets for downstream analyses or meta-analyses combining multiple
datasets.  In this example, we want to find datasets that are associated
with bipolar disorder, and we are only interested in human data.  In
addition, we will subset our results to datasets that have been batch
corrected.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">gemmapy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">api_instance</span> <span class="o">=</span> <span class="n">gemmapy</span><span class="o">.</span><span class="n">GemmaPy</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">api_response</span> <span class="o">=</span> <span class="n">api_instance</span><span class="o">.</span><span class="n">search_datasets</span><span class="p">([</span><span class="s2">&quot;bipolar&quot;</span><span class="p">],</span> <span class="n">taxon</span><span class="o">=</span><span class="s2">&quot;human&quot;</span><span class="p">,</span> <span class="n">limit</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">api_response</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>  <span class="c1"># view the object structure</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">api_response</span><span class="o">.</span><span class="n">data</span><span class="p">:</span>
<span class="gp">... </span>  <span class="k">if</span> <span class="n">d</span><span class="o">.</span><span class="n">geeq</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span>  <span class="n">d</span><span class="o">.</span><span class="n">geeq</span><span class="o">.</span><span class="n">batch_corrected</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">d</span><span class="o">.</span><span class="n">short_name</span><span class="p">,</span> <span class="n">d</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">d</span><span class="o">.</span><span class="n">bio_assay_count</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">GSE35974 Expression data from the human cerebellum brain 144</span>
<span class="go">GSE46416 State- and trait-specific gene expression in euthymia and mania 32</span>
</pre></div>
</div>
<p>We are left with two datasets. For simplicity, we’ll pick
<a class="reference external" href="https://gemma.msl.ubc.ca/expressionExperiment/showExpressionExperiment.html?id=8997">GSE46416</a>
since it has the smaller number of samples. Now that we have the ID
for our experiment, we can fetch the data associated with it.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">api_response</span> <span class="o">=</span> <span class="n">api_instance</span><span class="o">.</span><span class="n">get_datasets_by_ids</span><span class="p">([</span><span class="s2">&quot;GSE46416&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">api_response</span><span class="o">.</span><span class="n">data</span><span class="p">:</span>
<span class="gp">... </span>  <span class="nb">print</span><span class="p">(</span><span class="n">d</span><span class="o">.</span><span class="n">short_name</span><span class="p">,</span> <span class="n">d</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">d</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">GSE46416 State- and trait-specific gene expression in euthymia and mania 8997</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">d</span><span class="o">.</span><span class="n">description</span><span class="p">)</span>
<span class="go">Gene expression profiles of bipolar disorder (BD) patients were assessed during both a manic and a euthymic phase and compared both intra-individually, and with the gene expression profiles of controls.</span>
<span class="go">Last Updated (by provider): Sep 05 2014</span>
<span class="go">Contributors:  Christian C Witt Benedikt Brors Dilafruz Juraeva Jens Treutlein Carsten Sticht Stephanie H Witt Jana Strohmaier Helene Dukal Josef Frank Franziska Degenhardt Markus M Nöthen Sven Cichon Maren Lang Marcella Rietschel Sandra Meier Manuel Mattheisen</span>
</pre></div>
</div>
<p>To access the expression data in a convenient form, you can use
<a class="reference internal" href="api.html#gemmapy.GemmaPy.get_dataset_object" title="gemmapy.GemmaPy.get_dataset_object"><code class="xref py py-func docutils literal notranslate"><span class="pre">get_dataset_object()</span></code></a>. It is a high-level wrapper
that combines various endpoint calls to return an <a class="reference external" href="https://anndata.readthedocs.io/">anndata</a> (Annotated Data) object of the
queried dataset for downstream analyses. They include the expression
matrix along with the experimental design, and ensure the sample names
match between both when transforming/subsetting data.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span> <span class="o">=</span> <span class="n">api_instance</span><span class="o">.</span><span class="n">get_dataset_object</span><span class="p">(</span><span class="s2">&quot;GSE46416&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="go">AnnData object with n_obs × n_vars = 21986 × 32</span>
<span class="go">    obs: &#39;GeneSymbol&#39;, &#39;GeneName&#39;, &#39;NCBIid&#39;</span>
<span class="go">    var: &#39;batch&#39;, &#39;disease&#39;</span>
<span class="go">    uns: &#39;title&#39;, &#39;abstract&#39;, &#39;url&#39;, &#39;database&#39;, &#39;accession&#39;, &#39;GemmaQualityScore&#39;, &#39;GemmaSuitabilityScore&#39;, &#39;taxon&#39;</span>
</pre></div>
</div>
<p>To show how subsetting works, we’ll keep the manic phase data and the
<code class="code docutils literal notranslate"><span class="pre">reference_subject_role</span></code>s, which refers to the control samples in Gemma
datasets.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># Check the levels of the disease factor</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;disease&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
<span class="go">array([&#39;reference_subject_role&#39;, &#39;euthymic_phase_|_Bipolar_Disorder_|&#39;,</span>
<span class="go">       &#39;bipolar_disorder_|_manic_phase_|&#39;], dtype=object)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># Subset patients during manic phase and controls</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">manic</span><span class="o">=</span><span class="n">adata</span><span class="p">[:,(</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;disease&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;reference_subject_role&#39;</span><span class="p">)</span> <span class="o">|</span>
<span class="go">                      (adata.var[&#39;disease&#39;] == &#39;bipolar_disorder_|_manic_phase_|&#39;)].copy()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">manic</span><span class="p">)</span>
<span class="go">AnnData object with n_obs × n_vars = 21986 × 21</span>
<span class="go">    obs: &#39;GeneSymbol&#39;, &#39;GeneName&#39;, &#39;NCBIid&#39;</span>
<span class="go">    var: &#39;batch&#39;, &#39;disease&#39;</span>
<span class="go">    uns: &#39;title&#39;, &#39;abstract&#39;, &#39;url&#39;, &#39;database&#39;, &#39;accession&#39;, &#39;GemmaQualityScore&#39;, &#39;GemmaSuitabilityScore&#39;, &#39;taxon&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">manic</span><span class="o">.</span><span class="n">var</span><span class="p">)</span>
<span class="go">                                                 batch                           disease</span>
<span class="go">Control,1_DE50                       Batch_05_24/11/10            reference_subject_role</span>
<span class="go">Control,12                           Batch_02_26/11/09            reference_subject_role</span>
<span class="go">Control,9                            Batch_01_25/11/09            reference_subject_role</span>
<span class="go">Bipolardisorderpatientmanicphase,5   Batch_05_24/11/10  bipolar_disorder_|_manic_phase_|</span>
<span class="go">Control,15                           Batch_02_26/11/09            reference_subject_role</span>
<span class="go">Bipolardisorderpatientmanicphase,31  Batch_04_02/12/09  bipolar_disorder_|_manic_phase_|</span>
<span class="go">Bipolardisorderpatientmanicphase,29  Batch_03_27/11/09  bipolar_disorder_|_manic_phase_|</span>
<span class="go">Bipolardisorderpatientmanicphase,35  Batch_04_02/12/09  bipolar_disorder_|_manic_phase_|</span>
<span class="go">Bipolardisorderpatientmanicphase,18  Batch_02_26/11/09  bipolar_disorder_|_manic_phase_|</span>
<span class="go">Control,8                            Batch_01_25/11/09            reference_subject_role</span>
<span class="go">Control,3                            Batch_05_24/11/10            reference_subject_role</span>
<span class="go">Control,2_DE23                       Batch_05_24/11/10            reference_subject_role</span>
<span class="go">Control,2_DE40                       Batch_01_25/11/09            reference_subject_role</span>
<span class="go">Bipolardisorderpatientmanicphase,33  Batch_04_02/12/09  bipolar_disorder_|_manic_phase_|</span>
<span class="go">Control,4                            Batch_05_24/11/10            reference_subject_role</span>
<span class="go">Control,1_DE62                       Batch_01_25/11/09            reference_subject_role</span>
<span class="go">Bipolardisorderpatientmanicphase,10  Batch_01_25/11/09  bipolar_disorder_|_manic_phase_|</span>
<span class="go">Bipolardisorderpatientmanicphase,37  Batch_04_02/12/09  bipolar_disorder_|_manic_phase_|</span>
<span class="go">Bipolardisorderpatientmanicphase,23  Batch_03_27/11/09  bipolar_disorder_|_manic_phase_|</span>
<span class="go">Bipolardisorderpatientmanicphase,16  Batch_02_26/11/09  bipolar_disorder_|_manic_phase_|</span>
<span class="go">Bipolardisorderpatientmanicphase,21  Batch_03_27/11/09  bipolar_disorder_|_manic_phase_|</span>
</pre></div>
</div>
<p>Let’s check the expression for every sample to make sure they look OK:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># Plot Expression matrix</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">6</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">plt</span><span class="o">.</span><span class="n">boxplot</span><span class="p">(</span><span class="n">manic</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="n">sym</span><span class="o">=</span><span class="s1">&#39;.&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">([])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Samples&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Expression&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s1">&#39;ded.png&#39;</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/ded.png"><img alt="_images/ded.png" class="align-left" src="_images/ded.png" style="width: 100%;" /></a>
<p>Gene expression distributions of bipolar patients during manic phase and controls.</p>
<p>You can also use <a class="reference internal" href="api.html#gemmapy.GemmaPy.get_dataset_expression" title="gemmapy.GemmaPy.get_dataset_expression"><code class="xref py py-func docutils literal notranslate"><span class="pre">get_dataset_expression()</span></code></a> to only get the expression
matrix, and <a class="reference internal" href="api.html#gemmapy.GemmaPy.get_dataset_design" title="gemmapy.GemmaPy.get_dataset_design"><code class="xref py py-func docutils literal notranslate"><span class="pre">get_dataset_design()</span></code></a> to get the experimental design matrix.</p>
</section>
<section id="differential-expression-analyses">
<h2>Differential expression analyses<a class="headerlink" href="#differential-expression-analyses" title="Permalink to this heading">¶</a></h2>
<p>Gemma contains precomputed differential expression analyses for most
of its datasets. Analyses can involve more than one factor, such as
“sex” as well as “disease”. Some datasets contain more than one
analysis to account for different factors and their interactions. The
results are stored as resultSets, each corresponding to one factor (or
their interaction). You can access them using
<a class="reference internal" href="api.html#gemmapy.GemmaPy.get_differential_expression_values" title="gemmapy.GemmaPy.get_differential_expression_values"><code class="xref py py-func docutils literal notranslate"><span class="pre">get_differential_expression_values()</span></code></a>. From here on, we can
explore and visualize the data to find the most
differentially-expressed genes:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">gemmapy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pandas</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">api_instance</span> <span class="o">=</span> <span class="n">gemmapy</span><span class="o">.</span><span class="n">GemmaPy</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">de</span> <span class="o">=</span> <span class="n">api_instance</span><span class="o">.</span><span class="n">get_differential_expression_values</span><span class="p">(</span><span class="s1">&#39;GSE46416&#39;</span><span class="p">)</span>
<span class="go">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Classify probes for plotting</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">de</span><span class="p">[</span><span class="s1">&#39;diffexpr&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;No&#39;</span>   <span class="c1"># add extra column</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">de</span><span class="o">.</span><span class="n">loc</span><span class="p">[(</span><span class="n">de</span><span class="p">[</span><span class="s1">&#39;contrast_bipolar disorder, manic phase_logFoldChange&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mf">1.0</span><span class="p">)</span> <span class="o">&amp;</span>
<span class="gp">... </span>       <span class="p">(</span><span class="n">de</span><span class="p">[</span><span class="s1">&#39;contrast_bipolar disorder, manic phase_pvalue&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.05</span><span class="p">),</span><span class="s1">&#39;diffexpr&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Up&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">de</span><span class="o">.</span><span class="n">loc</span><span class="p">[(</span><span class="n">de</span><span class="p">[</span><span class="s1">&#39;contrast_bipolar disorder, manic phase_logFoldChange&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="o">-</span><span class="mf">1.0</span><span class="p">)</span> <span class="o">&amp;</span>
<span class="gp">... </span>       <span class="p">(</span><span class="n">de</span><span class="p">[</span><span class="s1">&#39;contrast_bipolar disorder, manic phase_pvalue&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.05</span><span class="p">),</span><span class="s1">&#39;diffexpr&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Down&#39;</span>
<span class="go">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Upregulated probes</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">de_up</span> <span class="o">=</span> <span class="n">de</span><span class="p">[</span><span class="n">de</span><span class="p">[</span><span class="s1">&#39;diffexpr&#39;</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;Up&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">de_up</span> <span class="o">=</span> <span class="n">de_up</span><span class="p">[[</span><span class="s1">&#39;Probe&#39;</span><span class="p">,</span><span class="s1">&#39;GeneSymbol&#39;</span><span class="p">,</span> <span class="s1">&#39;contrast_bipolar disorder, manic phase_pvalue&#39;</span><span class="p">,</span>
<span class="gp">... </span>        <span class="s1">&#39;contrast_bipolar disorder, manic phase_logFoldChange&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span>
<span class="gp">... </span>        <span class="s1">&#39;contrast_bipolar disorder, manic phase_pvalue&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">pandas</span><span class="o">.</span><span class="n">option_context</span><span class="p">(</span><span class="s1">&#39;display.max_rows&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="s1">&#39;display.max_columns&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">):</span>
<span class="gp">... </span>        <span class="nb">print</span><span class="p">(</span><span class="n">de_up</span><span class="p">[:</span><span class="mi">10</span><span class="p">])</span>
<span class="go">         Probe GeneSymbol  contrast_bipolar disorder, manic phase_pvalue  \</span>
<span class="go">11705  2319550       RBP7                                       0.000179</span>
<span class="go">13110  2548699     CYP1B1                                       0.000218</span>
<span class="go">11825  3907190       SLPI                                       0.000625</span>
<span class="go">18245  3629103      PCLAF                                       0.000961</span>
<span class="go">18386  3545525      SLIRP                                       0.001044</span>
<span class="go">3822   3146433      COX6C                                       0.001641</span>
<span class="go">12209  2538349        NaN                                       0.002130</span>
<span class="go">2873   2899102       H3C3                                       0.002146</span>
<span class="go">735    3635198     BCL2A1                                       0.002975</span>
<span class="go">11879  2633191      GPR15                                       0.003917</span>
<span class="go">       contrast_bipolar disorder, manic phase_logFoldChange</span>
<span class="go">11705                                              1.074</span>
<span class="go">13110                                              1.322</span>
<span class="go">11825                                              1.056</span>
<span class="go">18245                                              1.278</span>
<span class="go">18386                                              1.349</span>
<span class="go">3822                                               1.467</span>
<span class="go">12209                                              1.073</span>
<span class="go">2873                                               1.026</span>
<span class="go">735                                                1.080</span>
<span class="go">11879                                              1.205</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># Downregulated probes</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">de_dn</span> <span class="o">=</span> <span class="n">de</span><span class="p">[</span><span class="n">de</span><span class="p">[</span><span class="s1">&#39;diffexpr&#39;</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;Down&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">de_dn</span> <span class="o">=</span> <span class="n">de_dn</span><span class="p">[[</span><span class="s1">&#39;Probe&#39;</span><span class="p">,</span><span class="s1">&#39;GeneSymbol&#39;</span><span class="p">,</span> <span class="s1">&#39;contrast_bipolar disorder, manic phase_pvalue&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="s1">&#39;contrast_bipolar disorder, manic phase_logFoldChange&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="s1">&#39;contrast_bipolar disorder, manic phase_pvalue&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">pandas</span><span class="o">.</span><span class="n">option_context</span><span class="p">(</span><span class="s1">&#39;display.max_rows&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="s1">&#39;display.max_columns&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">):</span>
<span class="gp">... </span>        <span class="nb">print</span><span class="p">(</span><span class="n">de_dn</span><span class="p">[:</span><span class="mi">10</span><span class="p">])</span>
<span class="go">         Probe  GeneSymbol  contrast_bipolar disorder, manic phase_pvalue  \</span>
<span class="go">11215  2775390         NaN                                       0.000006</span>
<span class="go">15068  3760268         NaN                                       0.000031</span>
<span class="go">20918  3124344         NaN                                       0.000278</span>
<span class="go">4898   3673179         NaN                                       0.000313</span>
<span class="go">18551  3245871       WDFY4                                       0.000337</span>
<span class="go">2371   3022689    SND1-IT1                                       0.000448</span>
<span class="go">2294   2679014         NaN                                       0.000574</span>
<span class="go">15224  3336402       RBM14                                       0.000675</span>
<span class="go">15362  4019758         NaN                                       0.000685</span>
<span class="go">9709   3384417  ANKRD42-DT                                       0.000717</span>
<span class="go">       contrast_bipolar disorder, manic phase_logFoldChange</span>
<span class="go">11215                                             -1.556</span>
<span class="go">15068                                             -1.851</span>
<span class="go">20918                                             -1.037</span>
<span class="go">4898                                              -1.034</span>
<span class="go">18551                                             -1.157</span>
<span class="go">2371                                              -1.220</span>
<span class="go">2294                                              -1.175</span>
<span class="go">15224                                             -1.071</span>
<span class="go">15362                                             -1.405</span>
<span class="go">9709                                              -1.003</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># Add gene symbols as labels to DE genes</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">de</span><span class="p">[</span><span class="s1">&#39;delabel&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">de</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">de</span><span class="p">[</span><span class="s1">&#39;diffexpr&#39;</span><span class="p">]</span><span class="o">!=</span><span class="s1">&#39;No&#39;</span><span class="p">,</span><span class="s1">&#39;delabel&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">de</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">de</span><span class="p">[</span><span class="s1">&#39;diffexpr&#39;</span><span class="p">]</span><span class="o">!=</span><span class="s1">&#39;No&#39;</span><span class="p">,</span><span class="s1">&#39;GeneSymbol&#39;</span><span class="p">]</span>
<span class="go">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Volcano plot for bipolar patients vs controls</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">de</span><span class="p">[</span><span class="s1">&#39;-log10(p-value)&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">de</span><span class="p">[</span><span class="s1">&#39;contrast_bipolar disorder, manic phase_pvalue&#39;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">plotnine</span> <span class="kn">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">6</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">plot</span><span class="o">=</span><span class="p">(</span><span class="n">ggplot</span><span class="p">(</span><span class="n">de</span><span class="p">)</span>
<span class="gp">... </span><span class="o">+</span><span class="n">aes</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">x</span><span class="o">=</span><span class="s1">&#39;contrast_bipolar disorder, manic phase_logFoldChange&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">y</span><span class="o">=</span><span class="s1">&#39;-log10(p-value)&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">color</span><span class="o">=</span><span class="s1">&#39;diffexpr&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">labels</span><span class="o">=</span><span class="s1">&#39;delabel&#39;</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">... </span><span class="o">+</span><span class="n">geom_point</span><span class="p">()</span>
<span class="gp">... </span><span class="o">+</span><span class="n">geom_hline</span><span class="p">(</span><span class="n">yintercept</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="mf">0.05</span><span class="p">),</span> <span class="n">color</span> <span class="o">=</span> <span class="s2">&quot;gray&quot;</span><span class="p">,</span> <span class="n">linetype</span> <span class="o">=</span> <span class="s2">&quot;dashed&quot;</span><span class="p">)</span>
<span class="gp">... </span><span class="o">+</span><span class="n">geom_vline</span><span class="p">(</span><span class="n">xintercept</span> <span class="o">=</span> <span class="p">(</span><span class="o">-</span><span class="mf">1.0</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">),</span> <span class="n">color</span> <span class="o">=</span> <span class="s2">&quot;gray&quot;</span><span class="p">,</span> <span class="n">linetype</span> <span class="o">=</span> <span class="s2">&quot;dashed&quot;</span><span class="p">)</span>
<span class="gp">... </span><span class="o">+</span><span class="n">labs</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="s2">&quot;log2(FoldChange)&quot;</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="s2">&quot;-log10(p-value)&quot;</span><span class="p">)</span>
<span class="gp">... </span><span class="o">+</span><span class="n">scale_color_manual</span><span class="p">(</span><span class="n">values</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;blue&quot;</span><span class="p">,</span> <span class="s2">&quot;black&quot;</span><span class="p">,</span> <span class="s2">&quot;red&quot;</span><span class="p">))</span><span class="o">+</span><span class="n">theme_minimal</span><span class="p">())</span>
<span class="gp">...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">plot</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="s1">&#39;dea.png&#39;</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
<img alt="_images/dea.png" src="_images/dea.png" />
<p>Differentially-expressed genes in bipolar patients during manic phase versus controls.</p>
</section>
<section id="platform-annotations">
<h2>Platform Annotations<a class="headerlink" href="#platform-annotations" title="Permalink to this heading">¶</a></h2>
<p>Expression data in Gemma comes with annotations for the gene each
expression profile corresponds to. Using the
<a class="reference internal" href="api.html#gemmapy.GemmaPy.get_platform_annotations" title="gemmapy.GemmaPy.get_platform_annotations"><code class="xref py py-func docutils literal notranslate"><span class="pre">get_platform_annotations()</span></code></a> function, these
annotations can be retrieved independently of the expression data,
along with additional annotations such as Gene Ontology terms.</p>
<p>Examples:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">gemmapy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pandas</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">api_instance</span> <span class="o">=</span> <span class="n">gemmapy</span><span class="o">.</span><span class="n">GemmaPy</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">api_response</span> <span class="o">=</span> <span class="n">api_instance</span><span class="o">.</span><span class="n">get_platform_annotations</span><span class="p">(</span><span class="s1">&#39;GPL96&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">pandas</span><span class="o">.</span><span class="n">option_context</span><span class="p">(</span><span class="s1">&#39;display.max_rows&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="s1">&#39;display.max_columns&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">):</span> <span class="nb">print</span><span class="p">(</span><span class="n">api_response</span><span class="p">[:</span><span class="mi">6</span><span class="p">])</span>
<span class="go">     ProbeName    GeneSymbols                                      GeneNames  \</span>
<span class="go">0  211750_x_at  TUBA1A|TUBA1C              tubulin alpha 1a|tubulin alpha 1c</span>
<span class="go">1    216678_at            NaN                                            NaN</span>
<span class="go">2    216345_at         ZSWIM8             zinc finger SWIM-type containing 8</span>
<span class="go">3    207273_at            NaN                                            NaN</span>
<span class="go">4  216025_x_at         CYP2C9  cytochrome P450 family 2 subfamily C member 9</span>
<span class="go">5  218191_s_at         LMBRD1                      LMBR1 domain containing 1</span>
<span class="go">                                             GOTerms       GemmaIDs  \</span>
<span class="go">0  GO:0005737|GO:0000166|GO:0051234|GO:0005856|GO...  172797|360802</span>
<span class="go">1                                                NaN            NaN</span>
<span class="go">2  GO:0043170|GO:1990234|GO:0044260|GO:0050789|GO...         235733</span>
<span class="go">3                                                NaN            NaN</span>
<span class="go">4  GO:0005737|GO:0072330|GO:0008203|GO:0008202|GO...          32964</span>
<span class="go">5  GO:0043170|GO:0016192|GO:0051234|GO:0044260|GO...         303717</span>
<span class="go">      NCBIids</span>
<span class="go">0  7846|84790</span>
<span class="go">1         NaN</span>
<span class="go">2       23053</span>
<span class="go">3         NaN</span>
<span class="go">4        1559</span>
<span class="go">5       55788</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">api_response</span> <span class="o">=</span> <span class="n">api_instance</span><span class="o">.</span><span class="n">get_platform_annotations</span><span class="p">(</span><span class="s1">&#39;Generic_human&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">pandas</span><span class="o">.</span><span class="n">option_context</span><span class="p">(</span><span class="s1">&#39;display.max_rows&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="s1">&#39;display.max_columns&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">):</span> <span class="nb">print</span><span class="p">(</span><span class="n">api_response</span><span class="p">[:</span><span class="mi">6</span><span class="p">])</span>
<span class="go">      ProbeName   GeneSymbols  \</span>
<span class="go">0         LCN10         LCN10</span>
<span class="go">1      STAG3L5P      STAG3L5P</span>
<span class="go">2  LOC101059976  LOC101059976</span>
<span class="go">3          GAB3          GAB3</span>
<span class="go">4  LOC100287155  LOC100287155</span>
<span class="go">5        RASSF2        RASSF2</span>
<span class="go">                                           GeneNames  \</span>
<span class="go">0                                       lipocalin 10</span>
<span class="go">1                stromal antigen 3-like 5 pseudogene</span>
<span class="go">2  arf-GAP with GTPase, ANK repeat and PH domain-...</span>
<span class="go">3                  GRB2 associated binding protein 3</span>
<span class="go">4                  hypothetical protein LOC100287155</span>
<span class="go">5             Ras association domain family member 2</span>
<span class="go">                                             GOTerms GemmaIDs    NCBIids</span>
<span class="go">0        GO:0005576|GO:0005488|GO:0110165|GO:0036094   441399     414332</span>
<span class="go">1                                                NaN  8799043  101735302</span>
<span class="go">2                                                NaN  8779607  101059976</span>
<span class="go">3  GO:0002573|GO:0032502|GO:0030225|GO:0002521|GO...   389635     139716</span>
<span class="go">4                                                NaN  8090381  100287155</span>
<span class="go">5  GO:0048585|GO:0005737|GO:0043170|GO:0048584|GO...   201914       9770</span>
</pre></div>
</div>
<p>If you are interested in a particular gene, you can see which
platforms include it using
<a class="reference internal" href="api.html#gemmapy.GemmaPy.get_gene_probes" title="gemmapy.GemmaPy.get_gene_probes"><code class="xref py py-func docutils literal notranslate"><span class="pre">get_gene_probes()</span></code></a>. Note that functions to
search gene work best with unambigious identifiers rather than symbols.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># lists genes in gemma matching the symbol or identifier</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">api_response</span> <span class="o">=</span> <span class="n">api_instance</span><span class="o">.</span><span class="n">get_genes</span><span class="p">([</span><span class="s1">&#39;Eno2&#39;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">api_response</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="c1"># view the object structure</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">keys</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;official_symbol&#39;</span><span class="p">,</span><span class="s1">&#39;ensembl_id&#39;</span><span class="p">,</span><span class="s1">&#39;ncbi_id&#39;</span><span class="p">,</span><span class="s1">&#39;official_name&#39;</span><span class="p">,</span><span class="s1">&#39;taxon_common_name&#39;</span><span class="p">,</span><span class="s1">&#39;taxon_id&#39;</span><span class="p">,</span><span class="s1">&#39;taxon_scientific_name&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">api_response</span><span class="o">.</span><span class="n">data</span><span class="p">:</span> <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> </span><span class="si">%-18s</span><span class="s2"> </span><span class="si">%6d</span><span class="s2"> </span><span class="si">%-30s</span><span class="s2"> </span><span class="si">%-10s</span><span class="s2"> </span><span class="si">%2i</span><span class="s2"> </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">d</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()[</span><span class="n">k</span><span class="p">]</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">keys</span><span class="p">))</span>
<span class="gp">...</span>
<span class="go">ENO2 ENSG00000111674      2026 enolase 2                      human       1 Homo sapiens</span>
<span class="go">Eno2 ENSMUSG00000004267  13807 enolase 2, gamma neuronal      mouse       2 Mus musculus</span>
<span class="go">Eno2 ENSRNOG00000013141  24334 enolase 2                      rat         3 Rattus norvegicus</span>
<span class="go">ENO2 None               856579 phosphopyruvate hydratase ENO2 yeast      11 Saccharomyces cerevisiae</span>
<span class="go">eno2 ENSDARG00000014287 402874 enolase 2                      zebrafish  12 Danio rerio</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># ncbi id for human ENO2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">probs</span><span class="o">=</span><span class="n">api_instance</span><span class="o">.</span><span class="n">get_gene_probes</span><span class="p">(</span><span class="mi">2026</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">probs</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>  <span class="c1"># view the object structure</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># print only fields of interest</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">probs</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">6</span><span class="p">]:</span>
<span class="gp">... </span>  <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%-10s</span><span class="s2"> </span><span class="si">%-12s</span><span class="s2"> </span><span class="si">%-20s</span><span class="s2"> </span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span>
<span class="go">(d.name,d.array_design.short_name,d.array_design.name,d.array_design.taxon,</span>
<span class="go"> d.array_design.taxon_id,d.array_design.technology_type,d.array_design.troubled))</span>
<span class="go">20016      GPL3093      LC-25                human 1 TWOCOLOR False</span>
<span class="go">20024      GPL3092      LC-19                human 1 TWOCOLOR False</span>
<span class="go">20024      lymphochip-2 Lymphochip 37k       human 1 TWOCOLOR False</span>
<span class="go">1639       GPL962       CHUGAI 41K           human 1 TWOCOLOR False</span>
<span class="go">35850      NHGRI-6.5k   NHGRI-6.5k           human 1 TWOCOLOR False</span>
<span class="go">201313_at  GPL96        Affymetrix GeneChip Human Genome U133 Array Set HG-U133A human 1 ONECOLOR False</span>
</pre></div>
</div>
</section>
<section id="larger-queries">
<h2>Larger queries<a class="headerlink" href="#larger-queries" title="Permalink to this heading">¶</a></h2>
<p>Some endpoints accept multiple identifiers in a single
function call. For example, getting information on 2 datasets at the
same time.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">gemmapy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">api_instance</span> <span class="o">=</span> <span class="n">gemmapy</span><span class="o">.</span><span class="n">GemmaPy</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">api_response</span> <span class="o">=</span> <span class="n">api_instance</span><span class="o">.</span><span class="n">get_datasets_by_ids</span><span class="p">([</span><span class="s2">&quot;GSE35974&quot;</span><span class="p">,</span><span class="s2">&quot;GSE46416&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">api_response</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>  <span class="c1"># view the object structure</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">api_response</span><span class="o">.</span><span class="n">data</span><span class="p">:</span>
<span class="gp">... </span>            <span class="nb">print</span><span class="p">(</span><span class="n">d</span><span class="o">.</span><span class="n">short_name</span><span class="p">,</span> <span class="n">d</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">d</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">d</span><span class="o">.</span><span class="n">accession</span><span class="p">,</span> <span class="n">d</span><span class="o">.</span><span class="n">bio_assay_count</span><span class="p">,</span> <span class="n">d</span><span class="o">.</span><span class="n">taxon</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">GSE35974 Expression data from the human cerebellum brain 5939 GSE35974 144 human</span>
<span class="go">GSE46416 State- and trait-specific gene expression in euthymia and mania 8997 GSE46416 32 human</span>
</pre></div>
</div>
<p>To query large amounts of data, the API has a pagination system which
uses the <code class="code docutils literal notranslate"><span class="pre">limit</span></code> and <code class="code docutils literal notranslate"><span class="pre">offset</span></code> parameters. To avoid overloading the server,
calls are limited to a maximum of 100 entries, so the offset allows
you to get the next batch of entries in the next call(s). For
simplicity, this example shows how pagination works with 5 entries per
query.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">ofs</span> <span class="ow">in</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">10</span><span class="p">]:</span>
<span class="gp">... </span>    <span class="n">api_response</span><span class="o">=</span><span class="n">api_instance</span><span class="o">.</span><span class="n">get_platforms_by_ids</span><span class="p">([],</span><span class="n">offset</span><span class="o">=</span><span class="n">ofs</span><span class="p">,</span><span class="n">limit</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="gp">... </span>    <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">api_response</span><span class="o">.</span><span class="n">data</span><span class="p">:</span>
<span class="gp">... </span>        <span class="nb">print</span><span class="p">(</span><span class="n">d</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">d</span><span class="o">.</span><span class="n">short_name</span><span class="p">,</span> <span class="n">d</span><span class="o">.</span><span class="n">taxon</span><span class="p">)</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;--&#39;</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">1 GPL96 human</span>
<span class="go">2 GPL1355 rat</span>
<span class="go">3 GPL1261 mouse</span>
<span class="go">4 GPL570 human</span>
<span class="go">5 GPL81 mouse</span>
<span class="go">--</span>
<span class="go">6 GPL85 rat</span>
<span class="go">7 GPL339 mouse</span>
<span class="go">8 GPL91 human</span>
<span class="go">9 GPL890 rat</span>
<span class="go">10 GPL1406 mouse</span>
<span class="go">--</span>
<span class="go">11 GPL891 mouse</span>
<span class="go">12 GPL82 mouse</span>
<span class="go">13 GPL560 mouse</span>
<span class="go">14 GPL1073 mouse</span>
<span class="go">16 GPL369 mouse</span>
<span class="go">--</span>
</pre></div>
</div>
<p>The rest of the endpoints only support a single identifier:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">api_response</span> <span class="o">=</span> <span class="n">api_instance</span><span class="o">.</span><span class="n">get_dataset_annotations</span><span class="p">([</span><span class="s2">&quot;GSE35974&quot;</span><span class="p">,</span><span class="s2">&quot;GSE12649&quot;</span><span class="p">])</span>
<span class="go">...Error Traceback...</span>
</pre></div>
</div>
<p>In these cases, you will have to loop over all the identifiers you
wish to query and send separate requests:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">dataset</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;GSE35974&quot;</span><span class="p">,</span><span class="s2">&quot;GSE12649&quot;</span><span class="p">]:</span>
<span class="gp">... </span>    <span class="n">api_response</span> <span class="o">=</span> <span class="n">api_instance</span><span class="o">.</span><span class="n">get_dataset_annotations</span><span class="p">(</span><span class="n">dataset</span><span class="p">)</span>
<span class="gp">... </span>    <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">api_response</span><span class="o">.</span><span class="n">data</span><span class="p">:</span>
<span class="gp">... </span>        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">%s</span><span class="s1"> </span><span class="si">%-15s</span><span class="s1"> </span><span class="si">%-15s</span><span class="s1"> </span><span class="si">%-15s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">dataset</span><span class="p">,</span> <span class="n">d</span><span class="o">.</span><span class="n">object_class</span><span class="p">,</span> <span class="n">d</span><span class="o">.</span><span class="n">class_name</span><span class="p">,</span> <span class="n">d</span><span class="o">.</span><span class="n">term_name</span><span class="p">))</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;--&#39;</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">GSE35974 FactorValue     disease         Bipolar Disorder</span>
<span class="go">GSE35974 BioMaterial     biological sex  male</span>
<span class="go">GSE35974 BioMaterial     biological sex  female</span>
<span class="go">GSE35974 FactorValue     disease         schizophrenia</span>
<span class="go">GSE35974 ExperimentTag   organism part   cerebellum</span>
<span class="go">GSE35974 FactorValue     disease         mental depression</span>
<span class="go">--</span>
<span class="go">GSE12649 BioMaterial     organism part   reference subject role</span>
<span class="go">GSE12649 BioMaterial     disease         Bipolar Disorder</span>
<span class="go">GSE12649 ExperimentTag   organism part   prefrontal cortex</span>
<span class="go">GSE12649 BioMaterial     disease         schizophrenia</span>
<span class="go">--</span>
</pre></div>
</div>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">GemmaPy</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#about-gemma">About Gemma</a></li>
<li class="toctree-l2"><a class="reference internal" href="#installation-instructions">Installation instructions</a></li>
<li class="toctree-l2"><a class="reference internal" href="#additional-packages">Additional packages</a></li>
<li class="toctree-l2"><a class="reference internal" href="#downloading-expression-data">Downloading expression data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#differential-expression-analyses">Differential expression analyses</a></li>
<li class="toctree-l2"><a class="reference internal" href="#platform-annotations">Platform Annotations</a></li>
<li class="toctree-l2"><a class="reference internal" href="#larger-queries">Larger queries</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API Reference</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="index.html" title="previous chapter">gemmapy: a Python Wrapper for the Gemma API</a></li>
      <li>Next: <a href="api.html" title="next chapter">API Reference</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2022, The University of British Columbia.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 5.0.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/tutorial.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>